src/controller.js

```diff
@@ -0,0 +1,601 @@
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}?uploadId=${uploadId}`,
+      headers: {
+        'X-Amz-Content-Sha256': requestBodySha256.digest('hex'),
+        'Content-Length': requestBody.length,
+      },
+    });
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest),
+      body: requestBody,
+    });
+
+    let parsedResponse = parseS3Response(response.body);
+
+    let multipartEtag = this.__getMultipartEtag(parsedResponse, bucket, key);
+
+    if (response.statusCode !== 200) {
+      throw new Error('Could not complete a multipart upload');
+    }
+
+    if (tags) {
+      await this.__tagObject(opts);
+    }
+
+    return multipartEtag;
+  }
+
+  // http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadAbort.html
+  async abortMultipartUpload(opts) {
+   opts = runSchema(opts, Joi.object({
+     bucket: schemas.bucket.required(),
+     key: schemas.key.required(),
+     uploadId: schemas.uploadId.required(),
+   }));
+
+    let {bucket, key, uploadId} = opts;
+    let signedRequest = aws4.sign({
+      service: 's3',
+      region: this.region,
+      method: 'DELETE',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      headers: {
+        'content-length': 0,
+        'x-amx-content-sha256': emptysha256,
+      },
+      path: `/${key}?uploadId=`,
+    });
+
+    let response = await this.runner({
+      req: this.__serializeRequest(signedRequest),
+    });
+
+    parseS3Response(response.body);
+
+    if (response.statusCode !== 204) {
+      throw new Error('Could not abort multipart upload');
+    }
+  }
+
+  /**
+   * Generate the general request for uploading a resource to S3
+   * in a single request.
+   *
+   * NOTE: We still set the x-amz-meta-taskcluster-content-length because
+   * the multipart uploaded things must have this value set there as well.
+   *
+   * http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html
+   */
+  async generateSinglepartRequest(opts) {
+    opts = runSchema(opts, Joi.object().keys({
+      bucket: schemas.bucket.required(),
+      key: schemas.key.required(),
+      sha256: schemas.sha256.required(),
+      size: schemas.spSize.required(),
+      tags: schemas.tags,
+      permissions: schemas.permissions,
+    }).optionalKeys('tags', 'permissions'));
+
+    let {bucket, key, sha256, size, tags, permissions} = opts;
+
+    let unsignedRequest = {
+      service: 's3',
+      region: this.region,
+      method: 'PUT',
+      protocol: this.s3protocol,
+      hostname: this.__s3hostname(bucket),
+      path: `/${key}`,
+      headers: {
+        'content-length': size,
+        'x-amz-content-sha256': sha256,
+        'x-amz-meta-content-sha256': sha256,
+        'x-amz-meta-content-length': size,
+      }
+    };
+
+    // Set any tags.  For once, AWS Tags are atomic!
+    if (tags) {
+      unsignedRequest.headers['x-amz-tagging'] = qs.stringify(tags);
```

Comment:
For the record we just started using task.tags for something else too... Actions.json notably... So there is pros/cons... Maybe we have a way to select tags we want reflect... And we set those tags in queue config... Maybe...